### PR TITLE
py3.7/8/9 support, assorted minor fixes/refactor

### PR DESCRIPTION
**Pull request by justheuristic**
- removed match/case for py3.7/8/9 compatibility
- gather_op dim=-1 is no longer hard-coded into AllGather
- output_rules ALL is now 0, rationale: less things to worry about
- config "reduce" is now "sum"
- ran affected files through black / isort

### CODE REVIEW

File: petals_local_parallel/all_to_all_communication_primitives.py [hunk 1]
```diff
@@ -37,29 +44,30 @@ def __call__(self, x: torch.Tensor, rank: int):
...
+            if len(ranks_updated) == self.world_size:
+                parts_ready.set()  # can be called more than once; we dont care
             parts_ready.wait()
-            if self.ordered:
-                parts = sorted(parts)  # sorted by rank
-            parts = [part.to(x.device, non_blocking=True) for r, part in parts]
+            parts = [part.to(x.device, non_blocking=True) for part in parts]
             # note: for one of the parts with r == rank, part.to(device) is a no-op
-            return self.gather_op(parts, dim=-1)
+            return self.gather_op(parts)
```
Comment (justheuristic) [inline petals_local_parallel/all_to_all_communication_primitives.py:66]: rationale: AllGather is a generic op that can accept arbitrary gather_op callable; hardcoding dim here makes it hard-coded to torch.cat. Instead, we can set dim=-1 when instantiating AllGather (see slicer_wrapper.py diff)

**Review by BlackSamorez — APPROVED** [review]
lgtm
